// @azure/ai-form-recognizer/dist-esm/src/generated/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link StringIndexType} that the service accepts. */
export var KnownStringIndexType;
(function (KnownStringIndexType) {
    /** TextElements */
    KnownStringIndexType["TextElements"] = "textElements";
    /** UnicodeCodePoint */
    KnownStringIndexType["UnicodeCodePoint"] = "unicodeCodePoint";
    /** Utf16CodeUnit */
    KnownStringIndexType["Utf16CodeUnit"] = "utf16CodeUnit";
})(KnownStringIndexType || (KnownStringIndexType = {}));
/** Known values of {@link ApiVersion} that the service accepts. */
export var KnownApiVersion;
(function (KnownApiVersion) {
    /** TwoThousandTwentyTwo0831 */
    KnownApiVersion["TwoThousandTwentyTwo0831"] = "2022-08-31";
})(KnownApiVersion || (KnownApiVersion = {}));
/** Known values of {@link LengthUnit} that the service accepts. */
export var KnownLengthUnit;
(function (KnownLengthUnit) {
    /** Pixel */
    KnownLengthUnit["Pixel"] = "pixel";
    /** Inch */
    KnownLengthUnit["Inch"] = "inch";
})(KnownLengthUnit || (KnownLengthUnit = {}));
/** Known values of {@link SelectionMarkState} that the service accepts. */
export var KnownSelectionMarkState;
(function (KnownSelectionMarkState) {
    /** Selected */
    KnownSelectionMarkState["Selected"] = "selected";
    /** Unselected */
    KnownSelectionMarkState["Unselected"] = "unselected";
})(KnownSelectionMarkState || (KnownSelectionMarkState = {}));
/** Known values of {@link ParagraphRole} that the service accepts. */
export var KnownParagraphRole;
(function (KnownParagraphRole) {
    /** PageHeader */
    KnownParagraphRole["PageHeader"] = "pageHeader";
    /** PageFooter */
    KnownParagraphRole["PageFooter"] = "pageFooter";
    /** PageNumber */
    KnownParagraphRole["PageNumber"] = "pageNumber";
    /** Title */
    KnownParagraphRole["Title"] = "title";
    /** SectionHeading */
    KnownParagraphRole["SectionHeading"] = "sectionHeading";
    /** Footnote */
    KnownParagraphRole["Footnote"] = "footnote";
})(KnownParagraphRole || (KnownParagraphRole = {}));
/** Known values of {@link DocumentTableCellKind} that the service accepts. */
export var KnownDocumentTableCellKind;
(function (KnownDocumentTableCellKind) {
    /** Content */
    KnownDocumentTableCellKind["Content"] = "content";
    /** RowHeader */
    KnownDocumentTableCellKind["RowHeader"] = "rowHeader";
    /** ColumnHeader */
    KnownDocumentTableCellKind["ColumnHeader"] = "columnHeader";
    /** StubHead */
    KnownDocumentTableCellKind["StubHead"] = "stubHead";
    /** Description */
    KnownDocumentTableCellKind["Description"] = "description";
})(KnownDocumentTableCellKind || (KnownDocumentTableCellKind = {}));
/** Known values of {@link DocumentFieldType} that the service accepts. */
export var KnownDocumentFieldType;
(function (KnownDocumentFieldType) {
    /** String */
    KnownDocumentFieldType["String"] = "string";
    /** Date */
    KnownDocumentFieldType["Date"] = "date";
    /** Time */
    KnownDocumentFieldType["Time"] = "time";
    /** PhoneNumber */
    KnownDocumentFieldType["PhoneNumber"] = "phoneNumber";
    /** Number */
    KnownDocumentFieldType["Number"] = "number";
    /** Integer */
    KnownDocumentFieldType["Integer"] = "integer";
    /** SelectionMark */
    KnownDocumentFieldType["SelectionMark"] = "selectionMark";
    /** CountryRegion */
    KnownDocumentFieldType["CountryRegion"] = "countryRegion";
    /** Signature */
    KnownDocumentFieldType["Signature"] = "signature";
    /** Array */
    KnownDocumentFieldType["Array"] = "array";
    /** Object */
    KnownDocumentFieldType["Object"] = "object";
    /** Currency */
    KnownDocumentFieldType["Currency"] = "currency";
    /** Address */
    KnownDocumentFieldType["Address"] = "address";
})(KnownDocumentFieldType || (KnownDocumentFieldType = {}));
/** Known values of {@link DocumentSignatureType} that the service accepts. */
export var KnownDocumentSignatureType;
(function (KnownDocumentSignatureType) {
    /** Signed */
    KnownDocumentSignatureType["Signed"] = "signed";
    /** Unsigned */
    KnownDocumentSignatureType["Unsigned"] = "unsigned";
})(KnownDocumentSignatureType || (KnownDocumentSignatureType = {}));
/** Known values of {@link DocumentBuildMode} that the service accepts. */
export var KnownDocumentBuildMode;
(function (KnownDocumentBuildMode) {
    /** Template */
    KnownDocumentBuildMode["Template"] = "template";
    /** Neural */
    KnownDocumentBuildMode["Neural"] = "neural";
})(KnownDocumentBuildMode || (KnownDocumentBuildMode = {}));
/** Known values of {@link OperationKind} that the service accepts. */
export var KnownOperationKind;
(function (KnownOperationKind) {
    /** DocumentModelBuild */
    KnownOperationKind["DocumentModelBuild"] = "documentModelBuild";
    /** DocumentModelCompose */
    KnownOperationKind["DocumentModelCompose"] = "documentModelCompose";
    /** DocumentModelCopyTo */
    KnownOperationKind["DocumentModelCopyTo"] = "documentModelCopyTo";
})(KnownOperationKind || (KnownOperationKind = {}));
